velox/serializers/UnsafeRowSerde.cpp

```diff
@@ -0,0 +1,134 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/serializers/UnsafeRowSerde.h"
+#include "velox/row/UnsafeRowDeserializer.h"
+#include "velox/row/UnsafeRowDynamicSerializer.h"
+
+namespace facebook::velox::batch {
+
+std::optional<size_t> UnsafeRowKeySerializer::serializeKeys(
+    const std::shared_ptr<RowVector>& vector,
+    const std::vector<int>& keys,
+    vector_size_t row,
+    std::string_view& rowBuffer,
+    std::string_view& serializedKeys) {
+  char* buffer = const_cast<char*>(rowBuffer.data());
+  char* originalBuffer = buffer;
+  size_t keySize = 0;
+  buffer += sizeof(keySize);
+
+  auto fieldTypes =
+      std::dynamic_pointer_cast<const RowType>(vector->type())->children();
+  // Traverse the keys and serialize them
+  for (auto index : keys) {
+    if (index < 0 || index > fieldTypes.size()) {
+      return std::nullopt;
+    }
+    auto nextKeySize =
+        velox::row::UnsafeRowDynamicSerializer::serialize(
+            fieldTypes[index], vector->children()[index], buffer, row)
+            .value_or(0);
+    if (fieldTypes[index]->isFixedWidth()) {
+      nextKeySize = velox::row::UnsafeRow::kFieldWidthBytes;
+    }
+    // Moving buffer pointer forward and updating the size
+    keySize += nextKeySize;
+    buffer += nextKeySize;
+  }
+  // Writing the composite key size and buffer
+  *((size_t*)originalBuffer) = keySize;
+  serializedKeys = std::string_view(originalBuffer + sizeof(keySize), keySize);
+  return buffer - originalBuffer;
+}
+
+void UnsafeRowVectorSerde::initialize(
+    const std::shared_ptr<RowVector>& vector) {
+  velox::row::UnsafeRowDynamicSerializer::preloadVector(vector);
+}
+
+BatchSerdeStatus UnsafeRowVectorSerde::serializeRow(
+    const std::shared_ptr<RowVector>& vector,
+    const vector_size_t row,
+    std::string_view& serializedKeys,
+    std::string_view& serializedRow) {
+  // Get the row size without writing to the buffer
+  size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSizeRow(
+      vector->type(), vector.get(), row);
+
+  // Check if the current allocated buffer is large enough for this row
+  if (rowBuffer_.data() == nullptr || rowSize > rowBuffer_.size()) {
+    // Need more memory if we are serializing keys as well
+    // We allocate at least twice the needed size so if the new rows are
+    // bigger we do not have re-allocate
+    // In case there are keys we allocate three times the size
+    // The previously allocated buffer will be deallocated automatically
+    auto allocationSize = keys_.has_value() ? rowSize * 3 : rowSize * 2;
```

Comment:
sizeof 1 row seems a poor metric no, shouldnt it be probable size of all rows ?

Reply:
This function is only for writing one row to the output (like shuffle).

Reply:
What do you suggest? The buffer is supposed to hold one row at a time though. I can change the numbers to 10 and 5 if that's better.